improve: refactor walking for better responsivenes

nekiro opened the pull request:
# Description

More smooth, better, refactored walking

## Behavior

### **Actual**

Walking is laggy

### **Expected**

Should be smooth

## Fixes

Fix #1027

## Type of change

Please delete options that are not relevant.

  - [x] Bug fix (non-breaking change which fixes an issue)
  - [x] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [ ] This change requires a documentation update

## Checklist

  - [x] My code follows the style guidelines of this project
  - [x] I have performed a self-review of my own code
  - [ ] I checked the PR checks reports
  - [ ] I have commented my code, particularly in hard-to-understand areas
  - [ ] I have made corresponding changes to the documentation
  - [ ] My changes generate no new warnings
  - [ ] I have added tests that prove my fix is effective or that my feature works

nekiro committed:
fix: schedule next walk event for smoother turning

Zbizu reviewed (approved):
it's working, I had to set this one to minimum as well:
![obraz](https://github.com/user-attachments/assets/d62ad52b-ae59-42bb-8569-90c70e5dab40)

recording on otcr:

https://github.com/user-attachments/assets/ad064ef6-40c3-4f12-a2ee-dfe737abe5fd

kokekanon commented:
me test :
server : main repo otland
otcr:

pressing both keys (not holding down one key permanently)


https://github.com/user-attachments/assets/65e719b9-7842-4e14-a1d9-94c5ee16ad36

mehah commented:
@nekiro 
the feature that puts the delay on the first step has stopped working.
![image](https://github.com/user-attachments/assets/59844a64-91ea-4ed5-b5a2-319a89bfeb16)

because that was what reset the step when you didn't walk for 150ms.
![image](https://github.com/user-attachments/assets/f40c94c7-affe-4e68-84a5-e7002d39c65d)

nekiro committed:
Better walking

nekiro commented:
I have reworked the whole walking, it seems better to me. I also deleted some useless options.
It performs poorly on high ping though, so converting this into draft.

Zbizu commented:
I've added a cout at Game::playerMove in the server code to test this change
it appears that the client doesn't send the packet at all now
(tested with 12.x walk system)
no ping issues there, as the test was carried out on localhost server

https://github.com/user-attachments/assets/fe9f851a-dde4-480d-8c76-888c6a185a0b

nekiro commented:
> I've added a cout at Game::playerMove in the server code to test this change it appears that the client doesn't send the packet at all now (tested with 12.x walk system) no ping issues there, as the test was carried out on localhost server
> 
>  2025-01-12.06-49-44.mp4

I'm testing on tfs master and no such problems were spotted

BastardofWinterfell commented:
I gave this a try, but unfortunately, as is, at least for me, it does not **yet** solve the issues with walking. I've tried playing around with the delay settings in Options, but I am still having the same issues as before.

Here is a comparison between OTCv8 and this OTC of walking in a straight line and then suddenly stopping. I was pressing the W key to walk north and stopped pressing it as the char arrives at the gold coin. In OTC v8, the char stops exactly on the tile and immediately when I stopped pressing the W key, as it should. In this OTC, the char stops for a second on the correct tile, but then takes an extra step forward after I'm no longer pressing any key.

https://github.com/user-attachments/assets/43cf7f95-894c-48db-a4fb-25f3aaaedaee

Here is a comparison between OTCv8 and this OTC of walking back and forth between two specific tiles with gold coins to show how the clients handle **sudden direction changes**. In OTC v8, the character stops and turns around on the correct tile with the gold coin when I stop pressing one key and start pressing the other, as it should. In this OTC, the character seems to "stutter" during this direction change as if the client is predicting that I'm going to keep going instead of turning around.

https://github.com/user-attachments/assets/f33ba45d-1942-4a9c-bab7-ff939e4f79b2

Here is a comparison between OTCv8 and this OTC of walking over a path, marked by gold coins on the ground, and making two turns in different directions. In OTC v8, the character follows my exact key presses with no delay, no extra steps, no attempt to correct itself, and perfectly follows my intended path over the gold coins. In this OTC, the character "stutters" during the direction change and walks on the wrong tiles.

https://github.com/user-attachments/assets/1c1bb83d-99d5-45de-9f2b-c701690d8f90

Only difference in these videos is the client. I'm using the same character walking with the same speed, playing on the same server (**protocol 10.41**) with the same average ping (around 150 ms) and average fps (around 120) playing on the same PC.

I'm hoping this can eventually be sorted out and the walking system in this OTC made to behave and feel as fluid and **responsive** as OTCv8.

BastardofWinterfell commented:
After further investigation and testing, I was able to almost completely "fix" the walking issues I described in my previous comment with the following two changes:

1.  https://github.com/mehah/otclient/blob/main/src/client/localplayer.cpp#L44

Following @kokekanon's suggestion, changed this line from:
`return m_walkTimer.ticksElapsed() >= getStepDuration() && (dir == m_direction || !m_walking);`
to:
`return m_walkTimer.ticksElapsed() >= getStepDuration() && (!m_walking);`
This fixed the extra step the character was taking when suddenly stopping and the "stutter" when changing directions quickly.

2. https://github.com/mehah/otclient/blob/main/modules/game_walk/walk.lua#L41

Changed this line from:
`modules.game_interface.getRootPanel():setAutoRepeatDelay(200)`
to:
`modules.game_interface.getRootPanel():setAutoRepeatDelay(10)`
This fixed missed key presses which greatly increased the perceived responsiveness and fluidity of the character movement for me.

Not sure if this is how it is supposed to be, but it seems that the `AutoRepeatDelay` is being "set" in at least two different places in the code:
- https://github.com/mehah/otclient/blob/main/modules/game_walk/walk.lua#L41 (which was set to 200)
- https://github.com/mehah/otclient/blob/main/modules/game_interface/gameinterface.lua#L136 (which was set to 50)

nekiro commented:
> After further investigation and testing, I was able to almost completely "fix" the walking issues I described in my previous comment with the following two changes:
> 
> 1. https://github.com/mehah/otclient/blob/main/src/client/localplayer.cpp#L44
> 
> Following @kokekanon's suggestion, changed this line from: `return m_walkTimer.ticksElapsed() >= getStepDuration() && (dir == m_direction || !m_walking);` to: `return m_walkTimer.ticksElapsed() >= getStepDuration() && (!m_walking);` This fixed the extra step the character was taking when suddenly stopping and the "stutter" when changing directions quickly.
> 
> 2. https://github.com/mehah/otclient/blob/main/modules/game_walk/walk.lua#L41
> 
> Changed this line from: `modules.game_interface.getRootPanel():setAutoRepeatDelay(200)` to: `modules.game_interface.getRootPanel():setAutoRepeatDelay(10)` This fixed missed key presses which greatly increased the perceived responsiveness and fluidity of the character movement for me.
> 
> Not sure if this is how it is supposed to be, but it seems that the `AutoRepeatDelay` is being "set" in at least two different places in the code:
> 
> * https://github.com/mehah/otclient/blob/main/modules/game_walk/walk.lua#L41 (which was set to 200)
> * https://github.com/mehah/otclient/blob/main/modules/game_interface/gameinterface.lua#L136 (which was set to 50)

It was a draft (not yet done), not sure what about your fixes, but they seem to be workarounds. I did rework whole walking system just now, so let me know if this is now better than it was before.

One important thing to note for all other testers, you have to disable BOT_PROTECTION macro, because of the g_game.walk call from lua side. Explained it in #1037 

~~The only problem I noticed with current code is that diagonals arent properly timed, so character gets kinda stuck on trying to prewalk too fast. I'll take a look at this soon.~~

nekiro committed:
Fix diagonals incorrect prewalk

kokekanon commented:
simple test

| Server         	| Ping    	|   |
|----------------	|---------	|---	|
| tfs 8.6 (0.4)      	| 100-150 	|   ✅✅	(improved significantly)|
| tfs 1.5 (8.6)      	| 1       	|   	✅|
| tfs 1.4.2(10.98)   	| 1       	|   ✅	|
| tfs 1.6 (13.10)    	| 1       	|   ✅	|
| Canary (11.00) 	| 1       	|   	✅|
| Canary (13.40) 	| 1       	|   ✅	|

nekiro commented:
I will also take a look at how sudden speed changes behave with these changes, because before it was very buggy.

kokekanon commented:
Insignificant detail

lvl 73
sudden speed changes 
speed : 182 -> 40 (paralice)


## PR otcr 

_moonwalking_

https://github.com/user-attachments/assets/014abc79-155b-4915-87a0-4412e481dec7

vs
 
## v8

https://github.com/user-attachments/assets/811e4170-3479-4398-b076-5680454cb497

vs

## cipsoft 


https://github.com/user-attachments/assets/08cfc80b-03fe-4429-b87f-5a317da1b585

Zbizu commented:
## Environment

client: 13.20
server: tfs-based
platform: Windows for both
connection type: localhost
login type: protocol login (legacy charlist request over 7171 port)
tests are marked as OK when their results are same as rl client

## Test results

first step:
bot protection on: still broken ❌ (see #1037)
bot protection off: OK ✅

(rest of tests were taken with bot protection off)
two coins test - OK ✅
path of coins test - OK ✅
diagonal steps (numpad) - OK ✅
buffering next step - OK ✅ (pressing arrows to queue next step when walk locked on low speed)
stairhop/teleporthop test - OK ✅ - this test was about how fast I can move away from the stairs after I switched the floor - default setting was too high, but I was able to get 1:1 results when I set floor/teleport delay to 50ms
walk after turn - OK ✅ - **Even better than in rl client** (rl client has horrible delay for this, somewhere between 500 and 1000ms)
packets sent - OK ✅ - I no longer get disconnected for sending too many packets


Tests were repeated on Ubuntu dedicated server connecting with Windows client (ping: 30ms), same server engine. Results: OK (except for bot protection "on")
All tests were performed using OpenGL build

nekiro committed:
fix scrollbar option anchor

nekiro committed:
Add feature for prewalk and lower option delays for better experience

Zbizu commented on modules/game_walk/walk.lua:
```diff
@@ -183,19 +183,21 @@ function walk(dir)
+            end
 
-        player:lockWalk(200)
+            player:lockWalk(200)
```
should this be hardcoded?

nekiro replied:
Could be floorchange delay from options, true. I'll update it with next commit. I still have to fix speed change during prewalks.

BastardofWinterfell commented:
With these latest changes, I'm happy to inform that almost all issues I previously had with the walking being laggy and unresponsive have been solved. Therefore, first things first: **awesome work** @nekiro and thank you!

However, I am still getting an extra step after a key is no longer being pressed and stutters when changing directions quickly. Here's a video of what I mean:

https://github.com/user-attachments/assets/8ce3563d-2cda-4cba-a189-0ccb2e497d0d

If I implement this change to `localplayer.cpp#L44`, the extra step and stutter when changing directions no longer happens.

>  1. https://github.com/mehah/otclient/blob/main/src/client/localplayer.cpp#L44
>  
>  Following @kokekanon's suggestion, changed this line from: 
> `return m_walkTimer.ticksElapsed() >= getStepDuration() && (dir == m_direction || !m_walking);` to:
> `return m_walkTimer.ticksElapsed() >= getStepDuration() && (!m_walking);` 
> This fixed the extra step the character was taking when stopping and the "stutter" when changing directions quickly.

nekiro commented:
Does not reproduce for me on latest tfs master with 100-150 ping.

BastardofWinterfell commented:
I'm playing on a 10.41 protocol/client version TFS1.0-based server with 150~180 ping.
Not sure what causes the extra step and stutter, but with the change to `localplayer.cpp#L44`, it does not happen.
I figured I'd let you know in case there is a better solution somewhere else in the code.

nekiro commented:
> I'm playing on a 10.41 protocol/client version TFS1.0-based server with 150~180 ping. Not sure what causes the extra step and stutter, but with the change to `localplayer.cpp#L44`, it does not happen. I figured I'd let you know in case there is a better solution somewhere else in the code.

I'm no longer using the code you mention to change. This code was rewritten and looks different now. I tried with delay of 150-180 and still no walk back is happening for me. I'll try with program to delay my localhost network, so maybe I can reproduce it.

BastardofWinterfell commented:
My apologies, @nekiro. 
It was a noob mistake on my part.

I'm glad to report that after correctly pulling the latest changes from this PR, walking is working completely fine. No more issues, and controlling the character feels a million times better than before. No lag, no key press delay, no extra steps, no stutters when changing directions. All issues I was having with walking have been solved. Thank you very much.

mehah commented:
https://github.com/user-attachments/assets/c5258b11-cb5c-4a34-bf07-79c9de91fc9c

@nekiro 
I tested with Canary 13.40 and the camera is stuttering, I don't know if you can see it in the video, wait for someone else to confirm. (@kokekanon)


#### but if it's ok for everyone, we'll merge it without any problems.

nekiro commented:
> OTClient.-.Redemption.2025-01-14.18-40-53.mp4 
> @nekiro I tested with Canary 13.40 and the camera is stuttering, I don't know if you can see it in the video, wait for someone else to confirm. (@kokekanon)
> 
> #### but if it's ok for everyone, we'll merge it without any problems.

Can't see in video, but this pull request still need changes for sudden speed changes, so not ready to merge yet.

mehah commented:
https://github.com/user-attachments/assets/95f51f82-f658-4e07-b8a4-e4089dd3ee3f

I recorded it now with OBS, I think it's better, pay attention to the houses, the camera keeps shaking.

but in the video it is almost imperceptible.

kokekanon commented:
> I tested with Canary 13.40 and the camera is stuttering, I don't know if you can see it in the video, wait for someone

in my particular case 

Client: this PR GL.
Server : canary 13.40
PC: i7-10700k . 3080 nvidia. 32 gb ram

I don't notice the “camera is stuttering”.
I remember that a user once had it happen, he deactivated the v-sync and the camara stuttering was solved

mehah commented:
@nekiro @kokekanon 
resolved by setting ```g keyboard.setKeyDelay(key, 10)```, but I don't know what it might affect on other protocols.

![image](https://github.com/user-attachments/assets/ff886b13-8670-4494-852b-fd86aff1d8b3)

nekiro commented:
> @nekiro @kokekanon resolved by setting `g keyboard.setKeyDelay(key, 10)`, but I don't know what it might affect on other protocols.

It should be good, the only thing we can be afraid of is cpu usage with that fast repeat
I'll try to reproduce the stutter with current code and adjust accordingly.

mehah commented:
> > @nekiro @kokekanon resolved by setting `g keyboard.setKeyDelay(key, 10)`, but I don't know what it might affect on other protocols.
> 
> It should be good, the only thing we can be afraid of is cpu usage with that fast repeat I'll try to reproduce the stutter with current code and adjust accordingly.

setRepeat would only be for that key that the player would be pressing at that moment, when he released it, it would return to the normal state, so I believe it won't have much cost.

@nekiro 
in fact, we would have to put
```g keyboard.setKeyDelay(key, 10)``` inside g_keyboard.bindKeyDown
and ```g_keyboard.setKeyDelay(key, 30)``` inside g_keyboard.bindKeyUp

nekiro committed:
Change key repeat to 10ms

nekiro commented:
I believe this is ready to be merged. I did not work out the speed change bugs (they are the same as they were before these changes so it's okay), because I can't currently invest more time into it, so we can create a new issue for it (I think there is one already for it though #963).

rodrigopaixaorj commented:
@kokekanon @luanluciano93 @Zbizu @BastardofWinterfell 

Guys, can you test it?

mehah commented:
canary 13.40 locally, working as expected, now it will be necessary to test on other protocols.

rodrigopaixaorj committed:
Merge branch 'main' into fix-next-walk-event

BastardofWinterfell commented:
> @kokekanon @luanluciano93 @Zbizu @BastardofWinterfell
> 
> Guys, can you test it?

All working perfectly here. 
With these changes, walking is as responsive and fluid as it should be.

**Server/Protocol**: TFS-based 10.41

### Tested
✅ Walking in a straight line between two gold coins by holding movement key and stopping when key is unpressed.
✅ Walking precisely over a specific path with two turns in opposite directions (quick successive direction changes).
✅ Walking diagonally.
✅ Walking up/down stairs (stairhop/teleporthop).
✅ No lag, no delay, no stutters, no missed key presses, no extra steps.

nexsen commented:
TFS 1.4.2 - 10.98 - Testing, still have issues but when your speed has been modified for example by paralyze.
https://github.com/user-attachments/assets/5b47610e-20c7-431c-847a-48997f048ebe
